File: src/main/java/corp/soyummy/shoppinglists/dto/ShoppingListUpdateDto.java
package corp.soyummy.shoppinglists.dto;

import lombok.Getter;
import lombok.Setter;
import lombok.AllArgsConstructor;
import lombok.NoArgsConstructor;

import jakarta.validation.constraints.Size;
import jakarta.validation.constraints.Pattern;
import jakarta.validation.constraints.AssertTrue;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class ShoppingListUpdateDto {
    @Size(min = 1, message = "Ingredient must have at least one character.")
    private String strIngredient;

    @Pattern(regexp = "^\\d+$", message = "Weight must be a positive number")
    private String weight;

    @Size(min = 1, message = "Image must have at least one character.")
    private String image;

    @Size(min = 1, message = "Image must have at least one character.")
    private String recipeId;

    @AssertTrue(message = "At least one field (strIngredient, weight, image, recipeId) must be provided.")
    private boolean isAtLeastOneFieldProvided() {
        return (strIngredient != null && !strIngredient.isEmpty()) ||
                (weight != null && !weight.isEmpty()) ||
                (image != null && !image.isEmpty()) ||
                (recipeId != null && !recipeId.isEmpty());
    }
}

File: src/main/kotlin/com/soyummy/shoppinglists/dto/ShoppingListUpdateDto.kt
package com.soyummy.shoppinglists.dto

import jakarta.validation.constraints.Size
import jakarta.validation.constraints.Pattern
import jakarta.validation.constraints.AssertTrue

data class ShoppingListUpdateDto(
    @field:Size(min = 1, message = "Ingredient must have at least one character.")
    val strIngredient: String? = null,

    @field:Pattern(regexp = "^\\d+$", message = "Weight must be a positive number")
    val weight: String? = null,

    @field:Size(min = 1, message = "Image must have at least one character.")
    val image: String? = null,

    @field:Size(min = 1, message = "RecipeId must have at least one character.")
    val recipeId: String? = null
) {
    @AssertTrue(message = "At least one field (strIngredient, weight, image, recipeId) must be provided.")
    fun isAtLeastOneFieldProvided(): Boolean {
        return !strIngredient.isNullOrBlank() ||
                !weight.isNullOrBlank() ||
                !image.isNullOrBlank() ||
                !recipeId.isNullOrBlank()
    }
}